### PR TITLE
chore(Pubsub): Fix Avro samples (Publish and Subscribe)

### LINKED ISSUE #1812: Pubsub sample publish_avro_records is not working for Binary encoded data
This sample is throwing the error for AVRO Binary encoded message.

Sample: https://github.com/GoogleCloudPlatform/php-docs-samples/blob/main/pubsub/api/src/publish_avro_records.php
ErrorMessage:
"revisionInfo": "Could not validate message with any schema revision for schema: projects\/608760342633\/schemas\/samples-test-publish-avroBINARY-643d0861c4430 last checked revision_id=68017e67 failed with unknown status.",
                "message": "Message failed schema validation"

Issue can be reproduced by just running the testPublishAndSubscribeAvro (Inside SchemaTest.php) test.

### PR DESCRIPTION
**Pull request by ajupazhamayil**
Fixes #1812

### CODE REVIEW

**Comment by saranshdhingra** [comment]
@ajupazhamayil Is this due to a breaking change in a composer package?

**Comment by yash30201** [comment]
The way we're sending the avro data, it seems odd as the data is missing the schema. Generally Avro Data(optionally though), contains schemas for all the parts of data.  Like if we see the public datasets of google with Avro formats, we can see each of them contain a schema in the data from which rest of the data is deserialised.

@ajupazhamayil Can you please take a look into how can we incorporate the schema to the encoded avro data so that we can fetch the schema from the received data and de serialise them later([code pointer](https://github.com/GoogleCloudPlatform/php-docs-samples/pull/1781) for fetching avro data and de serialising them)?

**Comment by ajupazhamayil** [comment]
> @ajupazhamayil Is this due to a breaking change in a composer package?

Im note sure about this. the package `AvroDataIoWriter` started throwing error like message schema validation failed. Im not aware any package breaking changes.

**Comment by ajupazhamayil** [comment]
> The way we're sending the avro data, it seems odd as the data is missing the schema. Generally Avro Data(optionally though), contains schemas for all the parts of data. Like if we see the public datasets of google with Avro formats, we can see each of them contain a schema in the data from which rest of the data is deserialised.
> 
> @ajupazhamayil Can you please take a look into how can we incorporate the schema to the encoded avro data so that we can fetch the schema from the received data and de serialise them later([code pointer](https://github.com/GoogleCloudPlatform/php-docs-samples/pull/1781) for fetching avro data and de serialising them)?

Let me check this @yash30201  Thank you for the suggestion

**Comment by ajupazhamayil** [comment]
> @ajupazhamayil Is this due to a breaking change in a composer package?

@saranshdhingra I am unable to find the reason behind this issue. I tried even checking out last year commit to see this error is reproduced (Till Aug 3rd, this error is present). Still checking...

**Comment by ajupazhamayil** [comment]
> > @ajupazhamayil Is this due to a breaking change in a composer package?
> 
> @saranshdhingra I am unable to find the reason behind this issue. I tried even checking out last year commit to see this error is reproduced (Till Aug 3rd, this error is present). Still checking...

There was a change like this though: https://github.com/GoogleCloudPlatform/php-docs-samples/pull/1780/files

**Comment by ajupazhamayil** [comment]
> > > @ajupazhamayil Is this due to a breaking change in a composer package?
> > 
> > 
> > @saranshdhingra I am unable to find the reason behind this issue. I tried even checking out last year commit to see this error is reproduced (Till Aug 3rd, this error is present). Still checking...
> 
> There was a change like this though: https://github.com/GoogleCloudPlatform/php-docs-samples/pull/1780/files

But I don't think it is a cause of this issue.

File: pubsub/api/test/SchemaTest.php [hunk 1]
```diff
@@ -210,6 +210,7 @@ public function testPublishAndSubscribeAvro($encoding)
         $subscribeOutput = $this->runFunctionSnippet('subscribe_avro_records', [
```
Comment (vishwarajanand) [inline pubsub/api/test/SchemaTest.php:210]: Also add a test for `create_avro_schema`?

File: pubsub/api/src/publish_avro_records.php [hunk 2]
```diff
@@ -81,14 +81,10 @@ function publish_avro_records($projectId, $topicId, $definitionFile)
         $io = new AvroStringIO();
         $schema = AvroSchema::parse($definition);
         $writer = new AvroIODatumWriter($schema);
-        $dataWriter = new AvroDataIOWriter($io, $writer, $schema);
+        $encoder = new AvroIOBinaryEncoder($io);
+        $writer->write($messageData, $encoder);
 
-        $dataWriter->append($messageData);
```
[the remark this answers is not in the document]
Comment (vishwarajanand) [reply pubsub/api/src/publish_avro_records.php:86]: This approach works and it's consistent with other languages, I am okay with this route.
I can see that in `AvroDataIOWriter::__construct` we are writing the headers (here it means avro schema in some encoded format) so it appears this is not what we want. Instead `append` is neater way to go.

**Review by vishwarajanand — APPROVED** [review]

File: pubsub/api/test/SchemaTest.php (hunk 1 above)
Comment (ajupazhamayil) [reply pubsub/api/test/SchemaTest.php:210]: It is already available here: https://github.com/GoogleCloudPlatform/php-docs-samples/blob/main/pubsub/api/test/SchemaTest.php#L104

**Pull request merged**